[GLib][WebRTC] MDNS address registration support using Avahi

philn opened the pull request:
#### 2c02cea6d045a29d1f73dafc9b7e8e529ee6c6b8
<pre>
[GLib][WebRTC] MDNS address registration support using Avahi
<a href="https://bugs.webkit.org/show_bug.cgi?id=299005">https://bugs.webkit.org/show_bug.cgi?id=299005</a>

Reviewed by Michael Catanzaro.

Use DBus to register/unregister MDNS addresses of local network interfaces in Avahi. As this works,
enable it in WPE/GTK LibWebRTC builds.

* Source/WebKit/NetworkProcess/glib/NetworkMDNSRegisterGLib.cpp: Added.
(WebKit::NetworkMDNSRegister::NetworkMDNSRegister):
(WebKit::NetworkMDNSRegister::~NetworkMDNSRegister):
(WebKit::NetworkMDNSRegister::registerMDNSName):
* Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.cpp:
(WebKit::NetworkMDNSRegister::PendingRegistrationRequest::PendingRegistrationRequest):
(WebKit::pendingRegistrationRequestMap):
(WebKit::NetworkMDNSRegister::registerMDNSName):
(WebKit::NetworkMDNSRegister::unregisterMDNSNames):
(WebKit::PendingRegistrationRequest::PendingRegistrationRequest): Deleted.
* Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.h:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetworkManager.cpp:
(WebKit::LibWebRTCNetworkManager::GetMdnsResponder const):

Canonical link: <a href="https://commits.webkit.org/301397@main">https://commits.webkit.org/301397@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6ff0bebfb0320f3d4629df367d4c8a56b3776cfb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125792 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45454 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36205 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132660 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77668 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46137 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54013 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95827 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/63943 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128740 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36885 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112484 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76320 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35787 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76130 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106664 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30881 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135339 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52579 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40325 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104293 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53027 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108694 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104021 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26504 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49390 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27707 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/49918 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52474 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58283 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51822 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55171 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53516 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.cpp:
```diff
@@ -47,6 +47,14 @@ namespace WebKit {
+        static_cast<GDBusProxyFlags>(G_DBUS_PROXY_FLAGS_DO_NOT_CONNECT_SIGNALS | G_DBUS_PROXY_FLAGS_DO_NOT_LOAD_PROPERTIES), nullptr,
+        "org.freedesktop.Avahi", "/", "org.freedesktop.Avahi.Server", nullptr, &error.outPtr()));
+    if (error)
+        LOG_ERROR("Unable to connect to the Avahi daemon: %s", error->message);
```
A bit harsh? I think it's expected for avahi to not be running. I would either ignore the error here, or better: figure out which error is returned if avahi is not running, and print an error only if it's not that.

I'm not familiar with `LOG_ERROR()`; what exactly does that do? Is that release logging straight to the journal? I think we normally use `g_warning()` for stuff like this to ensure it appears in stderr.

philn replied:
Thanks for the review, I'll try to improve the patch next week :)

> I'm not familiar with LOG_ERROR(); what exactly does that do?

It expands to a `WTFReportError()` call which prints the msg to stderr.

mcatanzaro commented on Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.cpp:
```diff
@@ -192,15 +200,111 @@ void NetworkMDNSRegister::registerMDNSName(WebCore::ScriptExecutionContextIdenti
+#if USE(GLIB)
+    if (auto group = m_entryGroups.take(documentIdentifier)) [[likely]] {
+        g_dbus_proxy_call(group.get(), "Reset", g_variant_new("()"), G_DBUS_CALL_FLAGS_NONE, -1, nullptr, reinterpret_cast<GAsyncReadyCallback>(+[](GDBusProxy* proxy, GAsyncResult* result, gpointer userData) {
+            [[maybe_unused]] auto finalResult = adoptGRef(g_dbus_proxy_call_finish(proxy, result, nullptr));
```
Here I really _would_ create a `GUniqueOutPtr<Gerror>` and print a warning. It will probably never fail, but if it _does_ fail then it sure would be nice to have an error message.

mcatanzaro commented on Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.cpp:
```diff
@@ -192,15 +200,111 @@ void NetworkMDNSRegister::registerMDNSName(WebCore::ScriptExecutionContextIdenti
+    NetworkMDNSRegister::RegisterMDNSNameCallback callback;
+    std::optional<WebCore::ScriptExecutionContextIdentifier> documentIdentifier;
+};
+WEBKIT_DEFINE_ASYNC_DATA_STRUCT(EntryGroupData);
```
This would be better using regular old `std::unique_ptr` and `std::make_unique`. Then you can just `WTFMove()` the `unique_ptr` and stop worrying about the need to not forget to call to `destroyEntryGroupData()`.

mcatanzaro reviewed (dismissed):
So I _think_ what you've written here is actually currently safe. Consider adding a GCancellable (which NetworkMDNSRegister should cancel in its destructor) anyway:

 * Secondary benefit: some future developer (which might be you :) won't need to remember to do so if using the NetworkMDNSRegister or anything with lifetime tied to it in the D-Bus callbacks in the future. I see you don't do that today, though, so it's not required yet.
  * Primary benefit: you can cancel outstanding D-Bus calls immediately instead of waiting for a response. Checking for cancellation allows you to avoid printing errors with `LOG_ERROR` if the operation is canceled.
  
I think this is fine as-is the way you've written it today, but a cancellable would be cleaner and more robust to a future developer changing things.

mcatanzaro commented on Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.cpp:
```diff
@@ -47,9 +48,63 @@ namespace WebKit {
+        if (error) {
+            LOG_ERROR("Unable to connect to the Avahi daemon: %s", error->message);
+            return;
+        }
```
In all the async calls throughout this pull request (and, generally, anywhere), you don't want to print the error message on cancellation:

```suggestion
        if (error) {
                if (!g_error_matches(error.get(), G_IO_ERROR, G_IO_ERROR_CANCELLED))
                    LOG_ERROR("Unable to connect to the Avahi daemon: %s", error->message);
            return;
        }
```

philn replied:
Oh right, I think you mentioned it previously and I forgot about it... Thanks for the review, I'll try to update this soon.

mcatanzaro commented on Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.cpp:
```diff
@@ -47,9 +48,63 @@ namespace WebKit {
+NetworkMDNSRegister::~NetworkMDNSRegister()
+{
+#if USE(GLIB)
+    if (m_cancellable)
```
Surely it's guaranteed to be non-null?

(Even if not, `g_cancellable_cancel()` is actually NULL-safe anyway.)

mcatanzaro commented on Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.cpp:
```diff
@@ -47,9 +48,63 @@ namespace WebKit {
+    if (m_cancellable)
+        g_cancellable_cancel(m_cancellable.get());
+    for (auto& cancellable : pendingRegistrationCancellablesMap().values())
+        g_cancellable_cancel(cancellable.get());
```
Notice this is the only place where you cancel anything, so it should be possible to use just one cancellable and get rid of the map, right?

mcatanzaro commented on Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.cpp:
```diff
@@ -192,15 +229,106 @@ void NetworkMDNSRegister::registerMDNSName(WebCore::ScriptExecutionContextIdenti
+    auto identifier = PendingRegistrationRequestIdentifier::generate();
+    auto addResult = pendingRegistrationCancellablesMap().add(identifier, adoptGRef(g_cancellable_new()));
+    request->identifier = addResult.iterator->key;
+    request->cancellable = addResult.iterator->value;
```
For example, you could do:

```suggestion
    request->cancellable = m_cancellable;
```

mcatanzaro reviewed (commented):
This is enough platform-specific code that I suggest it's time to add NetworkMDNSRegisterGLib.cpp, so we don't have to add so much platform-specific code in the cross-platform file.

mcatanzaro commented on Source/WebKit/NetworkProcess/glib/NetworkMDNSRegisterGLib.cpp:
```diff
@@ -0,0 +1,153 @@
+        auto self = reinterpret_cast<NetworkMDNSRegister*>(userData);
+        self->m_dbusProxy = adoptGRef(g_dbus_proxy_new_for_bus_finish(result, &error.outPtr()));
+        if (error && !g_error_matches(error.get(), G_IO_ERROR, G_IO_ERROR_CANCELLED))
+            LOG_ERROR("Unable to connect to the Avahi daemon: %s", error->message);
```
I would do one more thing: check what error gets returned when the avahi service is unavailable, and avoid printing the warning in that case, because that's expected if avahi is simply not installed, or if we're running under Flatpak, for example. (Too bad we don't have a portal for this. I guess this functionality is just not going to work under Flatpak.)

philn replied:
This code runs in the network process, so it should work in Flatpak.

philn replied:
Well I haven't checked actually, I was assuming it'd be OK since that process isn't as tightly sandboxed as the WebProcess :)

philn replied:
BTW we'll most likely depend on this also in the GstWebRTC backend, if we want https://bugs.webkit.org/show_bug.cgi?id=238873 fixed.

mcatanzaro replied:
Flatpak apps don't have access to the host session bus. It's going to be dropped by xdg-dbus-proxy regardless of what process makes the D-Bus calls. I'm afraid we probably need a new portal.

(We could also allowlist the Avahi D-Bus API in Flatpak itself so that Flatpak apps can access it, but that only makes sense if it is designed to be used by malicious untrusted applications. I doubt it is?)

(Another option is to add sandbox holes in the application's flatpak-builder manifest. But I'm enforcing a "no new permissions" rule for Epiphany.)

philn replied:
OK... here's what I came up with...
```diff
diff --git a/Source/WebKit/NetworkProcess/glib/NetworkMDNSRegisterGLib.cpp b/Source/WebKit/NetworkProcess/glib/NetworkMDNSRegisterGLib.cpp
index 17171ef16c33..b4249c5beb41 100644
--- a/Source/WebKit/NetworkProcess/glib/NetworkMDNSRegisterGLib.cpp
+++ b/Source/WebKit/NetworkProcess/glib/NetworkMDNSRegisterGLib.cpp
@@ -48,7 +48,16 @@ NetworkMDNSRegister::NetworkMDNSRegister(NetworkConnectionToWebProcess& connecti
         GUniqueOutPtr<GError> error;
         auto self = reinterpret_cast<NetworkMDNSRegister*>(userData);
         self->m_dbusProxy = adoptGRef(g_dbus_proxy_new_for_bus_finish(result, &error.outPtr()));
-        if (error && !g_error_matches(error.get(), G_IO_ERROR, G_IO_ERROR_CANCELLED))
+        if (!error)
+            return;
+
+#if PLATFORM(GTK)
+        // Check if the connection to the system bus was refused, don't log an error when that
+        // happens because it is expected when running as a Flatpak app.
+        if (g_error_matches(error.get(), G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
+            return;
+#endif
+        if (!g_error_matches(error.get(), G_IO_ERROR, G_IO_ERROR_CANCELLED))
             LOG_ERROR("Unable to connect to the Avahi daemon: %s", error->message);
     }, this);
 }
@@ -89,7 +98,10 @@ void NetworkMDNSRegister::registerMDNSName(WebCore::ScriptExecutionContextIdenti
         GUniqueOutPtr<GError> error;
         auto variant = adoptGRef(g_dbus_proxy_call_finish(G_DBUS_PROXY(object), result, &error.outPtr()));
         if (error) {
-            if (!g_error_matches(error.get(), G_IO_ERROR, G_IO_ERROR_CANCELLED))
+            // We might have access to the system bus but Avahi is not activatable/installed, don't
+            // log an error when that is the case.
+            if (!g_error_matches(error.get(), G_IO_ERROR, G_IO_ERROR_CANCELLED) &&
+                !g_error_matches(error.get(), G_DBUS_ERROR, G_DBUS_ERROR_SERVICE_UNKNOWN))
                 LOG_ERROR("Unable to add Avahi entry group: %s", error->message);
             request->completionHandler(request->name, WebCore::MDNSRegisterError::Internal);
             return;
```

philn committed:
[GLib][WebRTC] MDNS address registration support using Avahi
https://bugs.webkit.org/show_bug.cgi?id=299005

Reviewed by Michael Catanzaro.

Use DBus to register/unregister MDNS addresses of local network interfaces in Avahi. As this works,
enable it in WPE/GTK LibWebRTC builds.

* Source/WebKit/NetworkProcess/glib/NetworkMDNSRegisterGLib.cpp: Added.
(WebKit::NetworkMDNSRegister::NetworkMDNSRegister):
(WebKit::NetworkMDNSRegister::~NetworkMDNSRegister):
(WebKit::NetworkMDNSRegister::registerMDNSName):
* Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.cpp:
(WebKit::NetworkMDNSRegister::PendingRegistrationRequest::PendingRegistrationRequest):
(WebKit::pendingRegistrationRequestMap):
(WebKit::NetworkMDNSRegister::registerMDNSName):
(WebKit::NetworkMDNSRegister::unregisterMDNSNames):
(WebKit::PendingRegistrationRequest::PendingRegistrationRequest): Deleted.
* Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.h:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetworkManager.cpp:
(WebKit::LibWebRTCNetworkManager::GetMdnsResponder const):

Canonical link: https://commits.webkit.org/301397@main
